### PR TITLE
Add new test. Verifies that VF is released after pod removal.

### PR DESCRIPTION
**Pull request by kononovn**
…oval

### CODE REVIEW

File: conformance/vfs_assignement_test.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+		BeforeEach(func() {
+			testNode = sriovInfos.Nodes[0]
+			intf, err := sriovInfos.FindOneSriovDevice(testNode)
+			Expect(err).ToNot(HaveOccurred(), "No SR-IOV supported devices detected")
+			Expect(intf.TotalVfs).To(BeNumerically(">", 2), "Cluster has less than 2 VFs available.")
+
+			nodePolicyConfig = &sriovv1.SriovNetworkNodePolicy{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "testpolicy",
```
Comment (fedepaol) [inline conformance/vfs_assignement_test.go]: Can you give the policy a more descriptive name (like vfreleasedpolicy or something)?

File: conformance/vfs_assignement_test.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+					NetworkNamespace: namespaces.Test,
+					Vlan:             0,
+					IPAM: `{ "type": "host-local","subnet": "10.10.10.0/24","rangeStart": 
+					"10.10.10.171","rangeEnd": "10.10.10.181","routes": [{ "dst": "0.0.0.0/0" }],"gateway": "10.10.10.1"}`,
+					ResourceName: nodePolicyConfig.Spec.ResourceName,
+				},
+			}
+			err = clients.Create(context.TODO(), networkConfig)
+			Expect(err).ToNot(HaveOccurred(), "Error to create SriovNetwork")
+			netAttDef := &netattdefv1.NetworkAttachmentDefinition{}
```
Comment (fedepaol) [inline conformance/vfs_assignement_test.go]: Can you use this inside the eventually block?

File: conformance/vfs_assignement_test.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+			Eventually(func() error {
+				return clients.Get(context.TODO(), runtimeclient.ObjectKey{Name: networkConfig.Name, Namespace: namespaces.Test}, netAttDef)
+			}, 30*time.Second, 1*time.Second).ShouldNot(HaveOccurred(), "Error to detect NetworkAttachmentDefinition")
+		})
+
+		Context("Virtual Functions", func() {
+			// 21396
+			It("The VFs should be released once the pod deleted and can be used by the new created pods", func() {
```
Comment (fedepaol) [inline conformance/vfs_assignement_test.go]: It (should release the vfs once ..) produce a more readable output

File: conformance/vfs_assignement_test.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+								Command: []string{"sleep", "10h"},
+							},
+						},
+					},
+				}
+
+				runningPodA, err := clients.Pods(testPodA.Namespace).Create(testPodA)
+				Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("Error to create pod %s", testPodA.Name))
+				By("Checking that first Pod is in Running state")
+				whaitForPodStatus(runningPodA, v1core.PodRunning, 3*time.Minute)
```
Comment (fedepaol) [inline conformance/vfs_assignement_test.go]: waitForPodStatus. Also, some coordination with https://github.com/openshift/sriov-tests/pull/18 will be required as both have a pod waiting logic. My vote goes to a local Eventually function, like we do here https://github.com/openshift/sriov-tests/pull/18/files#diff-c17e18d66aee5ed11707fbfcb5601f92R284

File: conformance/vfs_assignement_test.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+				err = clients.Pods(namespaces.Test).Delete(runningPodA.Name, &metav1.DeleteOptions{
+					GracePeriodSeconds: pointer.Int64Ptr(0),
+				})
+				Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("Error to delete pod %s", runningPodA.Name))
+				By("Checking that second pod is able to use released VF")
+				whaitForPodStatus(runningPodB, v1core.PodRunning, 3*time.Minute)
+			})
+
+			AfterEach(func() {
+				// Delete resources
```
Comment (fedepaol) [inline conformance/vfs_assignement_test.go]: This should be covered by `namespaces.Clean`

**Review by fedepaol — COMMENTED** [review]
looks good! First round of comments.

File: conformance/vfs_assignement_test.go (hunk 1 above)
Comment (kononovn) [reply conformance/vfs_assignement_test.go]: Done

File: conformance/vfs_assignement_test.go (hunk 3 above)
Comment (kononovn) [reply conformance/vfs_assignement_test.go]: I added Should release the VFs once the pod deleted and same VFs can be used by the new created pods. I hope it's more readable

**Comment by SchSeba** [comment]
Overall comment: @nikita133 please change the PR title and also the git commit message to be more clear

File: conformance/vfs_assignement_test.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	. "github.com/onsi/gomega/gstruct"
```
Comment (SchSeba) [inline conformance/vfs_assignement_test.go:10]: nit: move this to the gomega section please

File: conformance/vfs_assignement_test.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+	var _ = Describe("Configuration", func() {
+		var nodePolicyConfig *sriovv1.SriovNetworkNodePolicy
+		var networkConfig *sriovv1.SriovNetwork
+		var testNode string
+
+		BeforeEach(func() {
+			testNode = sriovInfos.Nodes[0]
+			intf, err := sriovInfos.FindOneSriovDevice(testNode)
+			Expect(err).ToNot(HaveOccurred(), "No SR-IOV supported devices detected")
+			Expect(intf.TotalVfs).To(BeNumerically(">=", 2), "Cluster has less than 2 VFs available.")
```
Comment (SchSeba) [inline conformance/vfs_assignement_test.go]: maybe change this to more and not less in the comment?

File: conformance/vfs_assignement_test.go [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+			Expect(err).ToNot(HaveOccurred(), "Error to create SriovNetwork")
+			Eventually(func() error {
+				netAttDef := &netattdefv1.NetworkAttachmentDefinition{}
+				return clients.Get(context.TODO(), runtimeclient.ObjectKey{Name: networkConfig.Name, Namespace: namespaces.Test}, netAttDef)
+			}, 30*time.Second, 1*time.Second).ShouldNot(HaveOccurred(), "Error to detect NetworkAttachmentDefinition")
+		})
+
+		Context("Virtual Functions", func() {
+			// 21396
+			FIt("should release the VFs once the pod deleted and same VFs can be used by the new created pods", func() {
```
Comment (SchSeba) [inline conformance/vfs_assignement_test.go]: remove the `F` from here please

**Review by SchSeba — CHANGES_REQUESTED** [review]
Great test I have to small nits.

One more comment I have is this test can be problematically related to the other tests we have.
Let me explain when we use a mellanox card the next test after this one will configure 5 VF and to add them the sriov-config-daemon will reboot the node.

I was thinking can you configure here 5 (and move the all 5) into the first pod this way we stay consistent for the other tests. Or we can change the network in the other ones to 2 (not sure if there is a test that needs more the 2 VF)

@fedepaol what do you think?

File: conformance/vfs_assignement_test.go (hunk 8 above)
Comment (SchSeba) [reply conformance/vfs_assignement_test.go]: The `F` is here :P

File: conformance/vfs_assignement_test.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+		var networkConfig *sriovv1.SriovNetwork
+		var testNode string
+
+		BeforeEach(func() {
+			testNode = sriovInfos.Nodes[0]
+			intf, err := sriovInfos.FindOneSriovDevice(testNode)
+			Expect(err).ToNot(HaveOccurred(), "No SR-IOV supported devices detected")
+			Expect(intf.TotalVfs).To(BeNumerically(">=", 5), "Cluster has less than 5 VFs available.")
+
+			nodePolicyConfig = &sriovv1.SriovNetworkNodePolicy{
```
Comment (SchSeba) [inline conformance/vfs_assignement_test.go]: please use this function `CreateSriovPolicy`

File: conformance/vfs_assignement_test.go [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+				})), "Error SriovNetworkNodeState doesn't conatin reqired elements")
+			Eventually(func() int64 {
+				testedNode, err := clients.Nodes().Get(testNode, metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				resNum, _ := testedNode.Status.Capacity["openshift.io/testresource"]
+				capacity, _ := resNum.AsInt64()
+				return capacity
+			}, 3*time.Minute, time.Second).Should(Equal(int64(5)), "Error discovered less than 5 VFs")
+
+			networkConfig = &sriovv1.SriovNetwork{
```
Comment (SchSeba) [inline conformance/vfs_assignement_test.go]: please use `CreateSriovNetwork`

File: conformance/vfs_assignement_test.go [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+			Expect(err).ToNot(HaveOccurred(), "Error to create SriovNetwork")
+			Eventually(func() error {
+				netAttDef := &netattdefv1.NetworkAttachmentDefinition{}
+				return clients.Get(context.TODO(), runtimeclient.ObjectKey{Name: networkConfig.Name, Namespace: namespaces.Test}, netAttDef)
+			}, 30*time.Second, 1*time.Second).ShouldNot(HaveOccurred(), "Error to detect NetworkAttachmentDefinition")
+		})
+
+		Context("Virtual Functions", func() {
+			// 21396
+			FIt("should release the VFs once the pod deleted and same VFs can be used by the new created pods", func() {
```
Comment (SchSeba) [inline conformance/vfs_assignement_test.go]: please remove the `F` from this test

File: conformance/vfs_assignement_test.go [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+				By("Create first Pod which consumes all available VFs")
+				testPodA := &v1core.Pod{
```
Comment (SchSeba) [inline conformance/vfs_assignement_test.go]: please use this function `pod.DefineWithNetworks([]string{mtuNetwork.Name})`

File: conformance/vfs_assignement_test.go [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+				Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("Error to create pod %s", testPodA.Name))
+				By("Checking that first Pod is in Running state")
+
+				Eventually(func() v1core.PodPhase {
+					runningPodA, err = clients.Pods(namespaces.Test).Get(runningPodA.Name, metav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					return runningPodA.Status.Phase
+				}, 3*time.Minute, time.Second).Should(Equal(v1core.PodRunning))
+				By("Create second Pod which consumes one more VF")
+				testPodB := &v1core.Pod{
```
Comment (SchSeba) [inline conformance/vfs_assignement_test.go]: please use this function pod.DefineWithNetworks([]string{mtuNetwork.Name})

File: conformance/vfs_assignement_test.go [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+				By("Checking that second pod is able to use released VF")
+				Eventually(func() v1core.PodPhase {
+					runningPodB, err = clients.Pods(namespaces.Test).Get(runningPodB.Name, metav1.GetOptions{})
+					Expect(err).ToNot(HaveOccurred())
+					return runningPodB.Status.Phase
+				}, 3*time.Minute, time.Second).Should(Equal(v1core.PodRunning))
+			})
+		})
+
+		AfterEach(func() {
```
Comment (SchSeba) [inline conformance/vfs_assignement_test.go]: please move this one up next to `BeforeAll`

**Review by SchSeba — CHANGES_REQUESTED** [review]
almost there I have a few comments

**Comment by fedepaol** [comment]
Looks fine after @SchSeba 's comments

File: conformance/vfs_assignement_test.go [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+				nodeState, err := clients.SriovNetworkNodeStates(operatorNamespace).Get(testNode, metav1.GetOptions{})
+				Expect(err).ToNot(HaveOccurred())
+				return nodeState.Spec.Interfaces
+			}, 1*time.Minute, 1*time.Second).Should(ContainElement(MatchFields(
+				IgnoreExtras,
+				Fields{
+					"Name":     Equal(intf.Name),
+					"NumVfs":   Equal(numVfs),
+					"VfGroups": ContainElement(sriovv1.VfGroup{ResourceName: resourceName, DeviceType: "netdevice", VfRange: "0-4"}),
+				})), "Error SriovNetworkNodeState doesn't conatin reqired elements")
```
Comment (fedepaol) [inline conformance/vfs_assignement_test.go]: Spelling mistakes here, can you fix them?

**Pull request merged**